### ISSUE TITLE
Performance improvements for Executor#post

### ISSUE THREAD

**Issue by headius** [ruby-concurrency/concurrent-ruby#280]
The `#post` method in Executor has a very large amount of overhead:

* Several closures, including one created *solely* to pass args through to the given block.
* Use of java_method on *every* call.
* Creation of `[Runnable.java_class]` array on every call.

The closures are especially heavy because they force us to run this method with a heap-based local variable scope, even though it's *only* used for passing args to task.

I have a few small patches that improve this situation. They're not suitable for a PR at the moment (quick one-off work) but they're there.

https://gist.github.com/headius/3807e0180e83ed7c98a2

This patch does the following:

* Uses a Runnable-implementing Job class to hold args + block, rather than creating another closure.
* Decorates the executors as they're created to have a submit_runnable method that always uses submit(Runnable).

I tested this with both SingleThreadExecutor and ThreadPoolExecutor using the following code (just timing total run time and no warmup):

```ruby
exe = Concurrent::ThreadPoolExecutor.new
t = Time.now
i = 0
while i < 1_000_000
  exe.post { }
  i+=1
end
puts Time.now - t
```

With my changes the run time goes from 29s to around 5s.

I can help you massage these into place, but the patch as written is just a one-off.

**Comment by jdantonio** [comment]
@headius Thank you very much for noticing this. I'm currently working on some updates to our executors. I'll take a look at your patch and if I have any questions I'll ask ASAP.

**Comment by headius** [comment]
It would worth doing a pass over the rest of the codebase looking for
closures in hot methods. They're a perf hit on all impls compared to
methods, and you can often proc'ify a method and get the same iterator
pattern for free.

- Charlie (mobile)

**Comment by jdantonio** [comment]
@headius Maybe I'm doing something wrong, but I get this error:

```
cannot convert instance of class org.jruby.RubyModule to class java.lang.Class
```

at

```
java_alias :submit_runnable, :submit, [java.lang.Runnable]
```

Thoughts?

**Comment by headius** [comment]
Ohh sorry...I found that was an issue in 1.7 after I submitted. Add
.java_class to the end of Runnable.

**Comment by headius** [comment]
I fixed the gist, but I do want to point out that that particular part I only did to two of the executors, not to all of them. If there's a better way to decorate them all, that would be great...but until then this will need to be done for each place where Executors factory is used to create a new Executor.

**Comment by jdantonio** [comment]
Fixed in #283
